attic/rfc_reduction.md

```diff
@@ -0,0 +1,69 @@
+
+```
+ Feature Name: `Forecasting Reduction Techniques`
+ Start Date: 2019-11-16
+ Target Date:
+ ```
+
+## Summary
+[summary]: #summary
+ 
+ - Lagged values of a timeseries can be used as features for a machine learning algorithm.
+ - Two major apporaches: rolling window regression and expanding window regression
+
+## Motivation
+[motivation]: #motivation
+
+Most machine learning algorithms are not designed for temporal data. To still use these methods, we have to put the data in the right shape for our algorithm.
+
+
+## Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+
+
+Basically there are two major approaches to feed timeseries data into a regression method: the "rolling window"-approach and the "expanding window"-approach. Both methods use the last k values as features to predict the next value(s) in the timeseries.
+
+Example for k = 2: Predict X(t) given X(t-1) and X(t-2)
+
+```
+   X(t) X(t-1) X(t-2)
+1     1     NA     NA
+2     2      1     NA
+3     3      2      1
+4     4      3      2
+5     5      4      3
+6     6      5      4
+7     7      6      5
+8     8      7      6
+9     9      8      7
+10   10      9      8
+
+```
+The difference between these two approaches is the amount of data that is used for making the prediction. While the rolling window (with window size n) approach uses the last n values for training and then predicts the next values, the expanding window approach trains the model before every prediction on all the available data.
+
+Example: Suppose we are at timestep 50 of a timeseries
+
+Rolling window approach:
+
+1. choose a window-size (for simplicity we choose 50)
+2. train the model on observation 1 to 50
+3. predict the value for timestep 51
+4. for the next prediction train the model on observation 2 to 51
```

Comment:
Do we really re-train? This seems very inefficient?
Additionally: Do we use `1 to 50` in step 4 aswell or only 2 to 51?

Reply:
For this approach we use timestep 2 to 51 in step 4. Assuming that a series has a seasonal component of length 50, I think it would only make sense to train the model with 50 / 100 / 150 / etc. timesteps.
The re-training seems indeed inefficient. We could of course use the model for more than one timestep. But as time goes by, the behavior of our timeseries may change and we need to adjust for this.

Reply:
Fixed window is sort of weird like that. In sort of a backwards way to me it relies on coefficient inconsistency across samples to make forecasts

Growing window is nice because you can just update most models with a new data point

Reply:
I am not 100% sure I understand everything perfectly.

I thought it would work as follows:
(For a fixed window of size 10 and 50 data points)

1. Each row is a sliding window, i.e. row 1 is 1:10, row 2 is 2:11, row 3 is 3:12 and so on.
2. The target for row 1 is 11, 12 for row 2 and so on.
3. We train a model on this data.
4. We predict step 51 from points 41 to 50, 52 from points 42 to 51(now predicted etc.)

Reply:
Yes exactly. 
I think a approach like this might be useful for cyclical data. Let X be a series with a cycle of length 12 (e.g. monthly data).  In my opinion it makes sense to only use full cycles for training. So when we observe  new values (e.g. new values for January and February) of the series and want to use them to update our model, we have to omit values (again values for January and February) at the beginning of our training set or else those months are 
overrepresented in the training set.

Reply:
@pfistfl lol sorry now I'm confused. Are you bunching 1:3 together and then doing predictions only in step 4?

Reply:
The graph on this page may help paint the picture, though it's for CV

http://topepo.github.io/caret/data-splitting.html#data-splitting-for-time-series

![image](https://user-images.githubusercontent.com/5857231/69261135-0af0bf00-0b8f-11ea-80e6-49fd642a633a.png)

Reply:
So this RFC is about reduction techniques for regression, not CV.
I was talking about an approach that is called **Rolling Windows-based Regression** in the post below.

https://medium.com/making-sense-of-data/time-series-next-value-prediction-using-regression-over-a-rolling-window-228f0acae363

Reply:
I think that author is making nonsense words. The 'rolling window based regression' seems no different than an AR(2) [autoregressive with 2 lags) model. Besides that article I have never heard someone use the term 'rolling windows based regression'. 

Why does making a lagged term count as a 'reduction technique'? You'll have the same number of observations after the lag with one column now having an NA. That's sort of a reduction in a way I guess?

Reply:
What I was referring to is what the authors in this [paper](https://www.researchgate.net/publication/236941795_Machine_Learning_Strategies_for_Time_Series_Forecasting) call "Supervised Learning Setting".
I guess the [sktime paper](https://arxiv.org/pdf/1909.07872.pdf) people call it `tabluar regression` or [reduction techniques](https://alan-turing-institute.github.io/sktime/examples/forecasting.html#Reduction-strategies), which allows them to use a random forest regressor on the lagged (AR2) values.

You are right in that if you use a linear model, this would basically reduce to an AR model :)

Reply:
Ah icic, yeah the verbage is very weird in this area. idk, personally I think of it more like a preprocessing step or like setting up the Data Generating Process. If I was going to call them anything I'd say something like `temporal transforms`. Since instead of manipulating on the cross-section it manipulates against the temporal component of the data. This also kind of flows well into like verbage like `spatio-temporal transforms`, `cross-sectional transforms`, etc.

Naming things is hard!